design/healcheck-endpoint.md

```diff
@@ -0,0 +1,159 @@
+# Healthcheck endpoint
+
+* **Status**: Draft #1
+* **JIRA**: [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+* **Observability JIRA**: [KEYCLOAK-8288](https://issues.jboss.org/browse/KEYCLOAK-8288)
+
+## Motivation
+
+Deploying both Keycloak and RHSSO on top of Kubernetes requires defining both Readiness Check and Liveness Check
+(see the [manual](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)).
+Currently, for Keycloak we use a very naive implementation, that calls `/auth/realms/master` URL.
+This approach has many drawbacks, here are a few examples:
+
+* We cannot discover database failures
+* We cannot discover if certain Identity Provider is up and running
+* Undertow subsystem is down when proceeding with migration
+
+In order to address these problems, we need provide a parametrized Healthcheck endpoint in the Server and use it in our
+container image as well as in the Operator.
+
+## Parameterizing the Healthcheck URL
+
+The new Health Endpoint need to support query parameters in the endpoint in the following form:
+
+    http://localhost:8080/auth/health?timeout=2s&fail-on=db,idp-google
+
+The following list of additional parameters will be supported:
+
+* `fail-on` - a set of conditions under which the probe fails. The initial implementation should cover:
+  * `db` - failing on database failures.
+  * `idp-<identity-provider-alias>` - failing when an IdP is not accessible.
+  * `user-federation-<id>` - failing is LDAP connection is down.
+* (optional) `timeout` - the request gets terminated if not performed within a timeout.
+The implementation needs to take special care of releasing the thread pool (if necessary).
+Subsequent calls that take longer than expected should not exhaust the thread pools inside Wildfly. If not specified,
+`10s` will be used as default (the same timeout as the one used in Kubernetes).
+
+## Implementation
+
+From the implementation perspective, there are two ways of implementing Healcheck calls within Keycloak:
+* Active - whenever we check if a database is down (this is just an example), we initiate a new connection and
+check if it succeeds.
+* Passive - we position the implementation based on Metrics components (see the Observability Design). We actively
```

Comment:
Can you link to "Observability Design"?

I'd like to read up on what's proposed there

Reply:
Of course! It's in the review: https://github.com/keycloak/keycloak-community/pull/177

Just to give you some more context - the Healthcheck is part of the Observability effort. Those two should be considered together.

Reply:
Let me resolve this conversation.